src/devices/Hcsr04/Hcsr04.cs

```diff
@@ -40,32 +42,31 @@ public Sonar(int triggerPin, int echoPin)
         /// <summary>
         /// Gets the current distance in cm.
         /// </summary>
-        public double GetDistance()
+        private double GetDistance()
         {
             _timer.Reset();
 
             // Trigger input for 10uS to start ranging
             // ref https://components101.com/sites/default/files/component_datasheet/HCSR04%20Datasheet.pdf
+            _controller.Write(_trigger, PinValue.Low);
+            Thread.Sleep(TimeSpan.FromMilliseconds(0.002));
             _controller.Write(_trigger, PinValue.High);
             Thread.Sleep(TimeSpan.FromMilliseconds(0.01));
             _controller.Write(_trigger, PinValue.Low);
             
-            _timer.Start();
-            
             while(_controller.Read(_echo) == PinValue.Low)
             {
             }
-
+            _timer.Start();
```

Comment:
we should measure time starting from when we end sending the "High" pulse, so previously was correct

Reply:
No, Time should be measured when _echo change to high level. That's why the original distance needs to -8.

Reply:
if you start measuring from the _echo being high then you measure the pulse width and not the time between sending and receiving. I'll double read the spec to see if I didn't misunderstand anything.

Reply:
@ZhangGaoxing reading: https://cdn.sparkfun.com/datasheets/Sensors/Proximity/HCSR04.pdf

Spec wording is really weird. First they say:

```
The Echo is a distance object that is pulse width and the range in proportion
```

then

```
You can calculate the range through the time interval between sending trigger signal and receiving echo signal
and give formula
range = high level time * velocity (340M/S) / 2
```

I guess both will work then and the 8 I was seeing was some factor multiplied by response time from the sensor and formula still worked since it included actual time. I think this should work as is. Let me check with my sensor to see if I get correct results.

As for the initial low time: We don't need to set it directly since it's already low but we should make sure that at least 60ms elapsed since last measuring (spec says `we suggest to use over 60ms measurement cycle, in order to prevent trigger signal to the echo signal`). So we should check Environment.TickCount after we are finished and another time when we start and if 60ms hasn't elapsed we should wait until it does

Reply:
@krwq like this? But I feel that it takes more time than write the _trigger LOW.
```C#
private int lastMeasurment = 0;

private double GetDistance()
{
    _timer.Reset();

    while (Math.Abs(Environment.TickCount - lastMeasurment) < 60)
    {
    }
    _controller.Write(_trigger, PinValue.High);
    Thread.Sleep(TimeSpan.FromMilliseconds(0.01));
    _controller.Write(_trigger, PinValue.Low);

    while (_controller.Read(_echo) == PinValue.Low)
    {
    }

    _timer.Start();
    while (_controller.Read(_echo) == PinValue.High)
    {
    }
    _timer.Stop();

    lastMeasurment = Environment.TickCount;

    TimeSpan elapsed = _timer.Elapsed;

    return elapsed.TotalMilliseconds / 2.0 * 34.3;
}
```

Reply:
@ZhangGaoxing _trigger LOW does not protect against echo. I will push to your branch with the code.

I've just tested the new measuring code and it works really accurate except for 0-4cm range where spec says it doesn't work well anyway :smile:

Reply:
Actually the code above looks correct - also noticed the issue which we can fix since we are on it: when there is timeout (no object) it just hangs